### ISSUE TITLE
[CT-1004] [Bug] `dbt ls` does not respect the `--output=json` option when no nodes matched

### ISSUE THREAD

**Issue by moltar** [dbt-labs/dbt-core#5626]
### Is this a new bug in dbt-core?

- [X] I believe this is a new bug in dbt-core
- [X] I have searched the existing issues, and I could not find an existing issue for this bug

### Current Behavior

Running the `dbt ls` command with the `--output=json` that results in no nodes being matched returns a non-JSON output.

### Expected Behavior

An empty array `[]`  should be returned.

### Steps To Reproduce

Run:

````
dbt ls --output=json --select="tag:does_not_exist_123"
````

### Relevant log output

```shell
11:12:39  The selection criterion 'tag:does_not_exist_123' does not match any nodes
11:12:39  No nodes selected!
```

### Environment

```markdown
- OS: macOS
- Python: 3.8
- dbt: 1.2.0
```

### Which database adapter are you using with dbt?

postgres

### Additional Context

N/A

**Comment by jtcohen6** [comment]
> An empty array `[]` should be returned.

Good point — that would make sense to me, too :)

**Comment by moltar** [comment]
I think I was editing the issue I nuked my point about the use case.

The use is being able to pipe the output to other tools, consistently. E.g. running the output thru `jq` or ingesting it with another tool.

**Comment by lostmygithubaccount** [comment]
@jtcohen6 I would suspect this is a larger issue beyond just `dbt ls`? do we have reasonable confidence that the CLI's outputs would conform to shell standards for additional scripting? 

if not, I would think it makes sense to track this along with the CLI/API work?

**Comment by dbeatty10** [comment]
## TL;DR - Recommended implementation

Just suppress the `warn_or_error` call when `--output json`:
https://github.com/dbt-labs/dbt-core/blob/43e3fc22c4eae4d3d901faba05e33c40f1f1dc5a/core/dbt/task/list.py#L71-L73

It will cause the output to be completely empty, which is probably what we want (rather than `[]`).

Read on for the nitty grittys!

**Comment by dbeatty10** [comment]
## Reproducing the report
To try to reproduce the use case, I first installed `jq` per [these](https://stedolan.github.io/jq/download/) instructions.

Then using the `zsh` shell and with the Jaffle Shop example project, this worked:
`dbt ls --output json | jq 'length'`

As reported by @moltar, this didn't:
`dbt ls --output json --select "tag:does_not_exist_123" | jq 'length'`

Local installation is not necessary to do simple testing -- see below for how to use the online tool at https://jqplay.org/.

## Implementation location

Here's where the nodes are being collected and dumped to JSON one at a time:
https://github.com/dbt-labs/dbt-core/blob/43e3fc22c4eae4d3d901faba05e33c40f1f1dc5a/core/dbt/task/list.py#L116-L128

A special case would need to be added to yield the string for an empty JSON array (`[]`) if no nodes were found.

Additionally, the `warn_or_error` message would need to be suppressed for this special case:
https://github.com/dbt-labs/dbt-core/blob/43e3fc22c4eae4d3d901faba05e33c40f1f1dc5a/core/dbt/task/list.py#L71-L73

## Key considerations

### JSON-L
`dbt ls` outputs [JSON-L](https://jsonlines.org/) like this (with lines truncated for readability):
```
{"resource_type": "model", "depends_on": {"macros": [], "nodes": ["model.jaffle_shop.stg_customers", ...
{"resource_type": "model", "depends_on": {"macros": [], "nodes": ["model.jaffle_shop.stg_orders", ... 
{"resource_type": "model", "depends_on": {"macros": [], "nodes": ["seed.jaffle_shop.raw_customers"] ...
```

The proposed behavior would yield a string like:
```
[]
```

So the empty case and the non-empty case would have different schemas. Not elegant very elegant, and I would assume would lead to downstream issues. 😢 

In this case, tools like `jq` still might have trouble 🤷 

Since `dbt ls --output json` is emitting JSON-L, the most elegant solution when no nodes are found would be to... emit nothing!

Can `jq` handle that? Yep!

#### Example 1
Single non-null input to `jq`:
```
$ echo '"fun"' | tail -n+1 | jq 'length'
3
```

☝️ This shows a length of 3 characters

#### Example 2
Null input to `jq`:
```
$ echo '"fun"' | tail -n+2 | jq 'length'
```
☝️ This shows that `jq` just didn't do anything because it didn't get anything.

## Testing input to `jq`
If `jq` is the exemplar downstream tool then https://jqplay.org/ is handy for testing different input against different `jq` filters. Just copy-n-paste the output from `dbt ls ...` into the "JSON" textarea of their website and specify the desired `jq` filter.

[jqplay.org example](https://jqplay.org/s/Z0_uCYxKerq)

<img width="678" alt="image" src="https://user-images.githubusercontent.com/44704949/184045286-4146ad27-ad04-434f-963a-483f51e50ea4.png">

**Comment by dbeatty10** [comment]
## Found while researching
### Deprecation warnings in stdout
During my local testing, I observed deprecation warnings going to standard out (immediately before the JSON-L).

So in that specific case, it will _definitely_ break anything like `jq` from reading the output since the deprecation warnings are not valid JSON.

Somehow silencing those deprecation warnings from hitting stdout for `dbt ls --output json` is probably desirable. I should also probably put it in a separate ticket. 😅

**Comment by dbeatty10** [comment]
@moltar could you review the rationale above and then confirm or deny if an empty string would work just as well (or better!) than `[]`?

**Comment by moltar** [comment]
If dbt is returning JSONL then I think blank is correct.

But then it should be documented as such, I think.

`jq` can deal with JSONL just fine.

```
❯ echo "{}\n{}" | jq
{}
{}
```

**Comment by jtcohen6** [comment]
> @jtcohen6 I would suspect this is a larger issue beyond just dbt ls? do we have reasonable confidence that the CLI's outputs would conform to shell standards for additional scripting?

`dbt ls` is really the only command where we explicitly expect / support users piping the output:
https://github.com/dbt-labs/dbt-core/blob/a3b018fd3b0c789c976e0e189788471edef8acb1/core/dbt/task/list.py#L55-L61

(Those TODOs were put there by yours truly)

Beyond that, I'd only expect users to be piping dbt output to `jq` in the sense of JSON-formatted structured logs. That config (`--log-format json`) is subtly different from the `--output json` config, which is totally unique to `ls`.

**Comment by dbeatty10** [comment]
Quick summary of the "accepted answer".

Do this:
```
$ dbt --quiet ls --output json
```

Instead of this:
```
$ dbt ls --output json
```
